dbms/src/Storages/KVStore/Decode/PartitionStreams.cpp

```diff
@@ -53,52 +54,24 @@ extern const int ILLFORMAT_RAFT_ROW;
 extern const int TABLE_IS_DROPPED;
 } // namespace ErrorCodes
 
-
-static DM::WriteResult writeRegionDataToStorage(
-    Context & context,
-    const RegionPtrWithBlock & region,
-    RegionDataReadInfoList & data_list_read,
-    const LoggerPtr & log)
+struct AtomicReadWriteCtx
 {
-    const auto & tmt = context.getTMTContext();
-    const auto keyspace_id = region->getKeyspaceID();
-    const auto table_id = region->getMappedTableID();
-    UInt64 region_decode_cost = -1, write_part_cost = -1;
-
-    DM::WriteResult write_result = std::nullopt;
-    /// Declare lambda of atomic read then write to call multiple times.
-    auto atomic_read_write = [&](bool force_decode) {
-        /// Get storage based on table ID.
-        auto storage = tmt.getStorages().get(keyspace_id, table_id);
-        if (storage == nullptr)
-        {
-            // - force_decode == false and storage not exist, let upper level sync schema and retry.
-            // - force_decode == true and storage not exist. It could be the RaftLog or Snapshot comes
-            //   after the schema is totally exceed the GC safepoint. And TiFlash know nothing about
-            //   the schema. We can only throw away those committed rows.
-            return force_decode;
-        }
-
-        /// Get a structure read lock throughout decode, during which schema must not change.
-        TableStructureLockHolder lock;
-        try
-        {
-            lock = storage->lockStructureForShare(getThreadNameAndID());
-        }
-        catch (DB::Exception & e)
-        {
-            // If the storage is physical dropped (but not removed from `ManagedStorages`) when we want to write raft data into it, consider the write done.
-            if (e.code() == ErrorCodes::TABLE_IS_DROPPED)
-                return true;
-            else
-                throw;
-        }
-
-        Block block;
-        bool need_decode = true;
-
-        // try to use block cache if exists
-        if (region.pre_decode_cache)
+    AtomicReadWriteCtx(
+        const LoggerPtr & log_,
+        const Context & context_,
+        const TMTContext & tmt_,
+        KeyspaceID keyspace_id_,
+        TableID table_id_)
+        : log(log_)
+        , context(context_)
+        , tmt(tmt_)
+        , keyspace_id(keyspace_id_)
+        , table_id(table_id_)
+    {}
+
+    std::optional<Block> tryUseDecodeCache(const RegionPtrWithBlock & region, ManageableStoragePtr & storage)
```

Comment:
The pre decode logic will be removed in the next pr.

Reply:
some tests rely on the Region with pre-decoded block. `GenRegionPreDecodeBlockData`

Reply:
Prehandling raft snapshot as Block is already deprecated. We can remove these tests IMO.